phoenix-core/src/it/java/org/apache/phoenix/query/MaxConcurrentConnectionsIT.java

```diff
@@ -0,0 +1,152 @@
+        assertEquals(0,GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue());
+
+        List<Connection> connections = new ArrayList<>();
+        String url = getUniqueUrl();
+
+        try{
+            for(int i = 0; i < maxConnections; i++) {
+                connections.add(DriverManager.getConnection(url, props));
+            }
+
+            try {
+                DriverManager.getConnection(url, props);
+                fail();
+            } catch(SQLException e) {
+                assertEquals(NEW_CONNECTION_THROTTLED.getErrorCode(),e.getErrorCode());
+                assertEquals(NEW_CONNECTION_THROTTLED.getSQLState(),e.getSQLState());
+            }
+
+        } finally {
+            for(Connection conn : connections) {
+                conn.close();
+            }
+        }
+    }
+
+    /**
+     * This tests the delete path which creates a internal phoenix connection per region
+     * @throws Exception
+     */
+    @Test
+    public void testDeleteRuntimeFailureClosesConnections() throws Exception {
+        String tableName = generateUniqueName();
+
+        //table with lots of regions
+        String ddl = "create table " + tableName +  "  (i integer not null primary key, j integer) SALT_BUCKETS=256 ";
+
+        Properties props = new Properties();
+        props.setProperty(QueryServices.ENABLE_SERVER_SIDE_DELETE_MUTATIONS,
+                String.valueOf(false));
+        props.setProperty(CLIENT_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(10));
+        props.setProperty(INTERNAL_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(10));
+
+        String deleteStmt = "DELETE FROM " + tableName + " WHERE 20 = j";
+
+        assertEquals(0,GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue());
+        assertEquals(0,GLOBAL_OPEN_INTERNAL_PHOENIX_CONNECTIONS.getMetric().getValue());
+        Connection conn = null;
+        try{
+            conn = DriverManager.getConnection(getUniqueUrl(), props);
+            try(Statement statement = conn.createStatement()) {
+                statement.execute(ddl);
+            }
+            //Enable delay for the delete
+            DelayedRegionServer.setDelayEnabled(true);
+            try(Statement statement = conn.createStatement()) {
+                statement.execute(deleteStmt);
+            }
+            fail();
+        } catch (SQLException e) {
+            assertEquals(NEW_CONNECTION_THROTTLED.getErrorCode(),e.getErrorCode());
+            assertEquals(NEW_CONNECTION_THROTTLED.getSQLState(),e.getSQLState());
+        } finally {
+            DelayedRegionServer.setDelayEnabled(false);
+            if(conn != null) {
+                conn.close();
+            }
+            long connections = GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue();
+            assertTrue(String.format("Found %d connections still open.",connections),2 >= connections);
+            connections = GLOBAL_OPEN_INTERNAL_PHOENIX_CONNECTIONS.getMetric().getValue();
```

Comment:
I'm not sure I understand this condition. Aren't we expecting 10 conns?

Reply:
No we are closing most of the connections now. The 2 was initially chosen as there were occasionally background tasks in phoenix making the assert inconsistent.

Reply:
Let's be careful not to create a flapper -- do we need a wait step for the background tasks to finish?

Reply:
Waits are by nature flappers as well unless carefully constructed with controls.  The system task table is the main one i'm concerned with.  Maybe there is a way to disable system.task for this integration test?

Reply:
I believe there is a way to change the task handling interval (set it to Long.MAX_VALUE or something for the sake of your tests). See https://github.com/apache/phoenix/blob/master/phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TaskRegionObserver.java#L67
FYI @kadirozde is there a cleaner way to do this?

Reply:
Spoke with Kadir offline, he also recommended setting the initial task to a high value.  I've incorporated both into my IT.